### PR TITLE
8341903: Implementation of Scoped Values (Fourth Preview)

### PR DESCRIPTION
**Pull request by theRealAph**
The fourth preview of scoped values.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8342205](https://bugs.openjdk.org/browse/JDK-8342205) to be approved
- [x] Commit message must refer to an issue
- [x] Change requires a JEP request to be targeted

### Issues
 * [JDK-8341903](https://bugs.openjdk.org/browse/JDK-8341903): Implementation of Scoped Values (Fourth Preview) (**Enhancement** - P4)
 * [JDK-8338456](https://bugs.openjdk.org/browse/JDK-8338456): JEP 487: Scoped Values (Fourth Preview) (**JEP**)
 * [JDK-8342205](https://bugs.openjdk.org/browse/JDK-8342205): Implementation of Scoped Values (Fourth Preview) (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21456/head:pull/21456` \
`$ git checkout pull/21456`

Update a local copy of the PR: \
`$ git checkout pull/21456` \
`$ git pull https://git.openjdk.org/jdk.git pull/21456/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21456`

View PR using the GUI difftool: \
`$ git pr show -t 21456`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21456.diff">https://git.openjdk.org/jdk/pull/21456.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21456#issuecomment-2405548816)
</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
@theRealAph The update to src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java needs to be included too.

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 1]
```diff
@@ -157,14 +157,19 @@
  * {@snippet lang=java :
  *     private static final ScopedValue<String> NAME = ScopedValue.newInstance();
 
- *     ScopedValue.runWhere(NAME, "duke", () -> {
- *         try (var scope = new StructuredTaskScope<String>()) {
+ *     ScopedValue.where(NAME, "duke").run(() -> {
+ *         // @link substring="open" target="StructuredTaskScope#open()" :
+ *         try (var scope = StructuredTaskScope.open()) {
  *
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/ScopedValue.java:163]: You'll need to revert this line as that is the updated STS API so will be update with its next JEP.

Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/ScopedValue.java:163]: Make sure to run jdk:tier1 and jdk:tier2 as that will run the tests that use this API.

**Comment by wangweij** [comment]
The one line change in `Subject.java` is fine.

/label remove security

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 2]
```diff
@@ -160,11 +160,15 @@
  *     ScopedValue.runWhere(NAME, "duke", () -> {
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/ScopedValue.java:160]: This should be `ScopedValue.where(NAME, "duke").run(() -> {`.

**Comment by AlanBateman** [comment]
The changes in update 90a344a3 looks okay, just one example in the ScopedValue docs that was missed (the example is correct in the loom repo).

**Comment by theRealAph** [comment]
Tier1 and Tier2 done. I had one tier2 failure, java/net/URL/EarlyOrDelayedParsing.java. I don't think that's anything to do with these changes

**Review by AlanBateman — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**